Update according Truffle team (#6)
rsksmart/rsk-react-express-box: README.md

(2) 598ccae5e4436fedeb0e798c0d254789c55a63401ebfc3ae8ddde29634ddfcde
(3) 09934b80f391e0024b8cb00cd73790fdf64c4d0509e144766414fee317cd3f4e
(4) ac745b84b6574b5738d364b43e0d471c9d5107504acc709c90f6f091b78c751b
(5) 449654cde095f2349113ef12a93e139b4302bc95adb3619d08adf53dde9b8847
(6) c217f12a89c352fc70b5f1bd5742314b4fb1bb1e35cb779fdb3c2390106355db
(7) 1d4c74dfa4e99e161130c18cc63938bb120a128cefbf1b9188efc678bf5722cb
(8) 0f44e0becf2e090db498a1b747d2a758fcc81fb0241f350d61117a9c6b1fa82e
(9) 85218c5eec657470dafeb09e6f7101f91d21bfe822fbeeecfc9275f798662a63

Mnemonic: virtual valve razor retreat either turn possible student grief engage attract fiber

⚠️  Important ⚠️  : This mnemonic was created for you by Truffle. It is not secure.
Ensure you do not use it on production blockchains, or else you risk losing funds.

truffle(develop)>
```

4. Take a look at the smart contract `SimpleStorage.sol`. You can check it out in folder `contracts`.

This smart contract has:

* A variable `storedData` to store a number
* A function `get()` to return the number stored at variable `storedData`
* A function `set()` to change the number stored at variable `storedData`

> 5. Compile and migrate the smart contract. Note inside the development console we don't preface commands with `truffle`.

Reviewer: ```suggestion
5. Compile and migrate the smart contract. Note, inside the development console we don't preface commands with `truffle`.
```

> To make sure you're in the development console, the command prompt must be `truffle(develop)>`

```shell
compile
```

The `compile output` should be similar to:

![truffle compile](/images/image-02.png)

```shell
migrate
```

And the `migrate output` should be similar to:

![truffle migrate](/images/image-03.png)

6. Running contract tests.

> Our box also comes with the file `TestSimpleStorage.js` for testing the smart contract. You can check it out in the `test` folder.

Reviewer: ```suggestion
Truffle box also comes with the file `TestSimpleStorage.js` for testing the smart contract. You can check it out in the `test` folder.
```

Author: No, I am not talking about "any" Truffle box, but this specific Truffle box.

Author: ```suggestion
This Truffle box also comes with the file `TestSimpleStorage.js` for testing the smart contract. You can check it out in the `test` folder.
```

Run this command in the development console:

```shell
test
```

This `test output` should be similar to:

![truffle test](/images/image-04.png)

Note the command varies slightly if you're in or outside of the development console.

```javascript
// inside the development console.
test

// outside the development console.
truffle test
```

## The application

Our box is a client-server: 
- Server uses Express JS to interact with the smart contract.
- Client uses React app.

> 7. Running in developer mode

Reviewer: ```suggestion
7. Running in development mode
```

> In another terminal (i.e. not in the truffle develop prompt), go to the `app` directory and run the app in developer mode.

Reviewer: ```suggestion
In another terminal (i.e. not in the truffle develop prompt), go to the `app` directory and run the app in development mode.
```

> Do not close the other terminal, which is running the Truffle development console, because it is our Blockchain simulator.
> 
> If you close it and then open it again, you need to deploy / migrate the smart contract again too!

```shell
cd app
npm run dev
```

> This command executes server and client applications and you can access in your browser:

Reviewer: ```suggestion
This command executes server and client applications and you can access it in your browser:
```

- Server: [http://localhost:8080/](http://localhost:8080/)
- Client: [http://localhost:3000/](http://localhost:3000/)

> Smart contract changes must be manually recompiled and migrated!

**NOTE**: This box is the starting point for the RSK tutorial [using rsk-react-express-box](https://developers.rsk.co/tutorials/truffle-boxes/rsk-react-express-box/).

8. Running the production server

> You can chhose to run only the production server. In another terminal (i.e. not in the truffle develop prompt), go to the `app` directory and run the app in production mode.

Reviewer: ```suggestion
You can choose to run only the production server. In another terminal (i.e. not in the truffle develop prompt), go to the `app` directory and run the app in production mode.
```

```shell
cd app
npm start
```

> If you have `yarn` installed, you can use `yarn start`.

Then go to your browser at [http://localhost:8080/](http://localhost:8080/)

![localhost port 8080](/images/image-11.png)

> When you are running only the production server, for any change you need to stop and run the server again for the updates to take effect!

9. To build the application for production, use the build script in the `app` folder. A production build will be in the `app/dist` folder.

```javascript
npm run build
```

Take a look in the `app/dist` folder:

![dist folder](/images/image-13.png)

## Using RSK networks

### Setup an account & get R-BTC

> - Get an address, learning how works the [account based RSK addresses](https://developers.rsk.co/rsk/architecture/account-based/ "Account based RSK addresses - RSK Developers Portal").

Reviewer: ```suggestion
- Get an address, using the [Account based RSK addresses](https://developers.rsk.co/rsk/architecture/account-based/ "Account based RSK addresses - RSK Developers Portal") to learn how it works.
```

Author: No, the page [account-based](https://developers.rsk.co/rsk/architecture/account-based/) didn't teach how to get an address. After you get an address (and I am not explain how to do it here), this page explains how the addresses work in RSK.

The **minimumGasPrice** is written in the block header by miners and establishes the minimum gas price that a transaction should have in order to be included in that block.

To update the **minimumGasPrice** in our project run this query using cURL:

**Testnet**

```shell
curl https://public-node.testnet.rsk.co/ -X POST -H "Content-Type: application/json" \
    --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest",false],"id":1}' \
    > .minimum-gas-price-testnet.json
```

**Mainnet**

```shell
curl https://public-node.rsk.co/ -X POST -H "Content-Type: application/json" \
    --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest",false],"id":1}' \
    > .minimum-gas-price-mainnet.json
```

This query saved the details of latest block to 
file .minimum-gas-price-testnet.json 
or .minimum-gas-price-mainnet.json, respectively.

In the `truffle-config.js`, we are reading the parameter `minimumGasPrice` in each json file.

> For more information about the **Gas** and **minimumGasPrice** please go to [gas page](https://developers.rsk.co/rsk/rbtc/gas/ "Gas - RSK Developers Portal").

Reviewer: ```suggestion
For more information about the **Gas** and **minimumGasPrice** please go to the [gas page](https://developers.rsk.co/rsk/rbtc/gas/ "Gas - RSK Developers Portal").
```

### Connect to RSK

1. Copy your mnemonic to `truffle-config.js`

```javascript
//Put your mnemonic here, be careful not to deploy your mnemonic into production!
const mnemonic = 'A_MNEMONIC';
```

Please be aware that we are using `HDWalletProvider` with RSK Networks derivations path:
- RSK Mainnet dpath: `m/44’/137’/0’/0`
- RSK Testnet dpath: `m/44’/37310’/0’/0`

For more information check [RSKIP57](https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP57.md).

2. Run the development console for any RSK network.

```shell
# Console for Testnet
truffle console --network testnet

# Console for Mainnet
truffle console --network mainnet
```

> 3. Migrate the smart contracts. We will do it running the below commands directly in the terminal, without using the truffle console now to show to you this alternative.

Reviewer: ```suggestion
3. Migrate the smart contracts. We will do it running the below commands directly in the terminal, without using the truffle console, this is to show you an alternative.
```

```shell
truffle migrate
```
4. Update Express JS component. 

The express server uses the [web3.js](https://web3js.readthedocs.io/) library to interact with the blockchain - writing code that reads and writes data from the blockchain with smart contracts.

> Choose which network you would like to connect the server to RSK Network 

Reviewer: ```suggestion
Choose which RSK network you would like to connect the server to 
```

and update line 11 of the file `app/src/server/index.js`

**Testnet**
```js    
const provider = new Web3.providers.HttpProvider("https://public-node.testnet.rsk.co");    
```

**Mainnet**
```js        
const provider = new Web3.providers.HttpProvider("https://public-node.rsk.co");
```

5. In a terminal, go to the `app` directory and run the app.

For example, this command will run the production server:

```shell
cd app
npm start
```

Then go to your browser at [http://localhost:8080/](http://localhost:8080/)

> Note that when you are connected to an RSK network, you do not need to leave open the Truffle console, because the app is connected via a public node, directly to the network.

## Next steps

- **Go to tutorial**

Go to the tutorial [using rsk-react-express-box](https://developers.rsk.co/tutorials/truffle-boxes/rsk-react-express-box/) to learn more about this project. We covered all the steps with more details, explanations, and images.

- **Find more documentation**

Check out the [RSK developers portal](https://developers.rsk.co/).

- **Do you have questions?**

> Ask in [RSK chat](https://gitter.im/rsksmart/getting-started).

Reviewer: ```suggestion
Ask in the [RSK Gitter Chat](https://gitter.im/rsksmart/getting-started).
```

Author: No, I wouldn't like to specify the name of the chat.

Author: ```suggestion
Ask in the [RSK chat](https://gitter.im/rsksmart/getting-started).
```
